### ISSUE TITLE
R4.1: switch default pool from LVM to BTRFS-Reflink

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#6476]
**The problem you're addressing (if any)**

In R4.0, the default install uses LVM thin pools.  However, LVM appears to be optimized for servers, which results in several shortcomings:

- Space exhaustion is handled poorly, requiring manual recovery.  This recovery may sometimes fail.
- It is not possible to shrink a thin pool.
- Thin pools slow down system startup and shutdown.

Additionally, LVM thin pools do not support checksums.  This can be achieved via dm-integrity, but that does not support TRIM.

**Describe the solution you'd like**

I propose that R4.1 use BTRFS+reflinks by default.

**Where is the value to a user, and who might that user be?**

BTRFS has checksums by default, and has full support for TRIM.  It is also possible to shrink a BTRFS pool without a full backup+restore.  BTRFS does not slow down system startup and shutdown, and does not corrupt data if metadata space is exhausted.

When combined with LUKS, BTRFS checksumming provides authentication: it is not possible to tamper with the on-disk data (except by rolling back to a previous version) without invalidating the checksum.  Therefore, this is a first step towards untrusted storage domains.  Furthermore, BTRFS is the default in Fedora 33 and openSUSE.

Finally, with BTRFS, VM images are just ordinary disk files, and the storage pool the same as the dom0 filesystem.  This means that issues like #6297 are impossible.

**Describe alternatives you've considered**

None that are currently practical.  bcachefs and ZFS are long-term potential alternatives, but the latter would need to be distributed as source and the former is not production-ready yet.

**Additional context**

I have had to recover manually from LVM thin pool problems (failure to activate, IIRC) on more than one occasion.  Additionally, the only supported interface to LVM is the CLI, which is rather clumsy.  The LVM pool requires nearly twice the amount of code as the BTRFS pool, for example.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

`man lvm`

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

#5053
#6297
#6184
#3244 (really a kernel bug)
#5826
#3230 ― since reflink files are ordinary disk files we could just rename them without needing a copy
#3964
everything in https://github.com/QubesOS/qubes-issues/search?q=lvm+thin+pool&state=open&type=issues

**Comment by iamahuman** [comment]
It might be a good idea to compare performance (seq read, rand read, allocation, overwrite, discard) between the three backends. See: #3639

**Comment by GWeck** [comment]
With regard to VM boot time, LVM storage pool was slightly faster than BTRFS, but this may be still within the margin of error (LVM: 7.43 s versus BTRFS: 8.15 s for starting a debian-10-minimal VM).

**Comment by DemiMarie** [comment]
Marking as RFC because this is by no means finalized.

**Comment by tlaurion** [comment]
@DemiMarie following [comment](https://github.com/tasket/wyng-backup/issues/75#issuecomment-820376379) I'm posting deconstructed thoughts here.  

No problem with QubesOS searching the best FS to switch for on 4.1 release, and questioning partition scheme, but i'm a bit lost on the direction of QubesOS 4.1 and the goals here (stability? performance? backups? portability? security?)

I was kind of against having dom0 having seperate LVM pool for space constrains resulting of the change, but agreed and accepted that the pool metadata exhaustion possibility was a real tangible issue that hit me a lot [before](https://github.com/QubesOS/qubes-issues/issues/5053#issuecomment-584712941), for which [resolution](https://groups.google.com/g/qubes-users/c/3r9MuyQHTUs/m/qLzKmpG4AQAJ) is sketchy and still not advertised in widget correctly for users simply upgrading and being hit with.
 
The [fix in new install](https://github.com/QubesOS/qubes-blivet/commit/933b35eba35cb33586a25b6f82f89ec5fe03f9d5) resolved the issue, while QubesOS  decided to split the dom0 pool out of main pool, so fixing pool issues on the system would be more easy for the end user or non existent. 

I am just not so sure why switching filesystem is on point now, where LVM thin provisioning seems to fit the goal, but willing to hear more about the advantages.

I am interested into the reasoning for such a switch, and the probabilities of doing so, since I am really interested into pushing wyng-backups farther, inside/outside of Heads inside/outside of QubesOS,  of grant/self funding the work so that QubesOS metadata would be included in wyng-backups, permitting  restore/verification/fresh deployment/revert from local(oem recovery VM)/remote source, just applying diff where required from ssh remote red only mountpoint.

This filesystem choice seems to be less relevant then what can make those [changes consume dom0 LVM](https://github.com/QubesOS/qubes-issues/issues/4371#issuecomment-761863958) which should be excludedof dom0 so that dmverity can be setuped under Heads/Safeboot.  But this is irrelevant to this ticket.

**Comment by DemiMarie** [comment]
> I am just not so sure why switching filesystem is on point now, where LVM thin provisioning seems to fit the goal, but willing to hear more about the advantages.

The advantages are listed above.  In short, a BTRFS pool is more flexible, and it offers possibilities (such as whole-system snapshots) that I do not believe are possible with LVM thin provisioning.  BTRFS also offers flexible quotas, and can always recover from out of space conditions provided that a small amount of additional storage (such as a spare partition set aside for the purpose) is available.  Furthermore, BTRFS checksumming and scrubbing appear to be useful.  Finally, new storage can be added to and removed from a BTRFS pool at any time, and the pool can be shrunk as well.

BTRFS also has disadvantages: its throughput is worse than LVM, and there are reports of bad performance on I/O heavy workloads such as QubesOS.  Benchmarks and user feedback will be needed to determine which is better, which is why this is an RFC.

> I am interested into the reasoning for such a switch, and the probabilities of doing so, since I am really interested into pushing wyng-backups farther, inside/outside of Heads inside/outside of QubesOS, of grant/self funding the work so that QubesOS metadata would be included in wyng-backups, permitting restore/verification/fresh deployment/revert from local(oem recovery VM)/remote source, just applying diff where required from ssh remote red only mountpoint.

I believe that `btrfs send` and `btrfs receive` offer the same functionality as wyng-backups, but am not certain as I never used either.  As far as the probability: this is currently only a proposal, and I am honestly not sure if switching this close to the R4.1 release date is a good idea.  In any case, LVM will continue to be fully supported ― this just flips the default in the installer.

**Comment by tasket** [comment]
@DemiMarie There are many questions swirling around advanced storage on Linux, but I think the main ones applicable here are about reliability and performance. Btrfs and Thin LVM appear to offer trade-offs on those qualities, and I don't think its necessarily a good move to switch the Qubes default for a slower storage scheme at this point; storage speed is critical for Qubes' usability and large disk image files with random write patterns are Btrfs' weakest point.

Running out of space is probably Thin LVM's weakest point, although this can be pretty easily avoided. For one, dom0 root is moving to a dedicated pool in R4.1, which will keep admin working in most situations. Adding more protections to the domU pool can also be done with some pretty simple userland code. (For those who are skeptical, note that this is the general approach taken by [Stratis.](https://lwn.net/Articles/755454/))

The above mentioned Btrfs checksums is a nice-to-have feature against accidental damage, but it unfortunately does not come close to providing authentication. To my knowledge, no CRC mode can do that even if its encrypted. Any attacker able to induce some calculated change in an encrypted volume would probably find the malleability of encrypted CRCs to be little or no obstacle. IMHO, the authentication aspect of the proposal is a non-starter. (BTW, it looks like _dm-integrity_ may be able to do this now along with `discard` support, if its journal mode supports internal tags.)

As for backups, Wyng basically exists because tools like `btrfs send` are constrained to using the same back end (Btrfs with admin privileges) which severely narrows the user's options for backup destinations. Wyng can also be adapted to any storage source that can create snapshots and report their deltas (Btrfs included).

The storage field also continues to evolve in interesting ways: Red Hat is creating Stratis while hardware manufacturers implemented NVMe objects and enhanced parallelism. Stratis appears to be based on none other than Thin LVM's main components (dm-thin, etc) in addition to dm-integrity, with XFS on top; all the layers are tied together to respond cohesively from a single management interface. This is being developed to avoid Btrfs maintenance and performance pitfalls.

I think some examination of Btrfs development culture may also be in order, as it has driven Red Hat to exasperation and a decision to drop Btrfs. I'm not sure just what it is about accepting Btrfs patches that presents a problem, but it makes me concerned that too much trust has been eroded and that Btrfs may become a casualty in 'storage wars' between an IBM / Red Hat camp and what I'd call an Oracle-centric camp.

---

FWIW, I was one of the first users to show how Qubes could take advantage of Btrfs reflinks for cloning and to request specific reflink support. Back in 2014, it was easy to assume Btrfs shortcomings would be addressed fairly soon, since those issues were so obvious. Yet they are still unresolved today.

My advice at this point is to wait and see – and experiment. There is an unfortunate dearth of comparison tests configured in a way that makes sense; they usually compare Btrfs to bare Ext4, for example, and almost always overlook LVM thin pools. So its mostly apples vs oranges. However, what little benchmarking I've seen of thin LVM suggests a performance advantage vs Btrfs that would be too large to ignore. There are also Btrfs modes of use we should explore, such as any performance gain from disabling CoW on disk images; if this were deemed desirable then the Qubes Btrfs driver would have to be refactored to use subvolume snapshots instead of reflinks. An XFS reflink comparison on Qubes would also be very interesting!

**Comment by DemiMarie** [comment]
> The above mentioned Btrfs checksums is a nice-to-have feature against accidental damage, but it unfortunately does not come close to providing authentication. To my knowledge, no CRC mode can do that even if its encrypted. Any attacker able to induce some calculated change in an encrypted volume would probably find the malleability of encrypted CRCs to be little or no obstacle. IMHO, the authentication aspect of the proposal is a non-starter. (BTW, it looks like _dm-integrity_ may be able to do this now along with `discard` support, if its journal mode supports internal tags.)

The way XTS works is that any change (by an attacker who does not have the key) will completely scramble a 128-bit block; my understanding is that a CRC32 with a scrambled block will only pass with probability 2⁻³².  That said, BTRFS also supports Blake2b and SHA256, which would be better choices.

**Comment by DemiMarie** [comment]
> @DemiMarie There are many questions swirling around advanced storage on Linux, but I think the main ones applicable here are about reliability and performance. Btrfs and Thin LVM appear to offer trade-offs on those qualities, and I don't think its necessarily a good move to switch the Qubes default for a slower storage scheme at this point; storage speed is critical for Qubes' usability and large disk image files with random write patterns are Btrfs' weakest point.

In retrospect, I agree.  That said (as you yourself mention below) XFS also supports reflinks and lacks this problem.

> Running out of space is probably Thin LVM's weakest point, although this can be pretty easily avoided. For one, dom0 root is moving to a dedicated pool in R4.1, which will keep admin working in most situations. Adding more protections to the domU pool can also be done with some pretty simple userland code. (For those who are skeptical, note that this is the general approach taken by [Stratis.](https://lwn.net/Articles/755454/))

Will it be possible to reserve space for use by discards?  A user needs to be able to free up space even if they make a mistake and let the pool fill up.

> The above mentioned Btrfs checksums is a nice-to-have feature against accidental damage, but it unfortunately does not come close to providing authentication. To my knowledge, no CRC mode can do that even if its encrypted. Any attacker able to induce some calculated change in an encrypted volume would probably find the malleability of encrypted CRCs to be little or no obstacle. IMHO, the authentication aspect of the proposal is a non-starter. (BTW, it looks like _dm-integrity_ may be able to do this now along with `discard` support, if its journal mode supports internal tags.)

The way XTS works is that any change (by an attacker who does not have the key) will completely scramble a 128-bit block; my understanding is that a CRC32 with a scrambled block will only pass with probability 2⁻³². That said, BTRFS also supports Blake2b and SHA256, which would be better choices.

> As for backups, Wyng basically exists because tools like `btrfs send` are constrained to using the same back end (Btrfs with admin privileges) which severely narrows the user's options for backup destinations. Wyng can also be adapted to any storage source that can create snapshots and report their deltas (Btrfs included).

Good to know, thanks!

> The storage field also continues to evolve in interesting ways: Red Hat is creating Stratis while hardware manufacturers implemented NVMe objects and enhanced parallelism. Stratis appears to be based on none other than Thin LVM's main components (dm-thin, etc) in addition to dm-integrity, with XFS on top; all the layers are tied together to respond cohesively from a single management interface. This is being developed to avoid Btrfs maintenance and performance pitfalls.
> 
> I think some examination of Btrfs development culture may also be in order, as it has driven Red Hat to exasperation and a decision to drop Btrfs. I'm not sure just what it is about accepting Btrfs patches that presents a problem, but it makes me concerned that too much trust has been eroded and that Btrfs may become a casualty in 'storage wars' between an IBM / Red Hat camp and what I'd call an Oracle-centric camp.

My understanding (which admittedly comes from a comment on Y Combinator) is that BTRFS moves _too fast_ to be used in RHEL.  RHEL is stuck on one kernel for an entire release, and rebasing BTRFS every release became too difficult, especially since Red Hat has no BTRFS developers.

> FWIW, I was one of the first users to show how Qubes could take advantage of Btrfs reflinks for cloning and to request specific reflink support. Back in 2014, it was easy to assume Btrfs shortcomings would be addressed fairly soon, since those issues were so obvious. Yet they are still unresolved today.
> 
> ---
> 
> My advice at this point is to wait and see – and experiment. There is an unfortunate dearth of comparison tests configured in a way that makes sense; they usually compare Btrfs to bare Ext4, for example, and almost always overlook LVM thin pools. So its mostly apples vs oranges. However, what little benchmarking I've seen of thin LVM suggests a performance advantage vs Btrfs that would be too large to ignore. There are also Btrfs modes of use we should explore, such as any performance gain from disabling CoW on disk images; if this were deemed desirable then the Qubes Btrfs driver would have to be refactored to use subvolume snapshots instead of reflinks. An XFS reflink comparison on Qubes would also be very interesting!

That it would be, especially when combined with Stratis.  The other major problem with LVM2 (and possibly dm-thin) seems to be snapshot and discard speeds; I expect XFS reflinks to mitigate most of those problems.

**Comment by tasket** [comment]
Ah, new Btrfs feature... Great! I'd consider enabling one of its hashing modes as being able to support authentication.

I'd still consider the Stratis concept to be more interesting for now, as Qubes' current volume management is pretty similar but potentially even better and simpler due to having a privileged VM environment.

**Comment by DemiMarie** [comment]
> Ah, new Btrfs feature... Great! I'd consider enabling one of its hashing modes as being able to support authentication.

Agreed.  While I am not aware of any way to tamper with a LUKS partition without invalidating a CRC, Blake2b is by far the better choice.

> I'd still consider the Stratis concept to be more interesting for now, as Qubes' current volume management is pretty similar but potentially even better and simpler due to having a privileged VM environment.

I agree, with one caveat: my understanding is that LUKS/AES-XTS-512 + BTRFS/Blake2b-256 is sufficient to protect against even malicious block devices, whereas `dm-integrity` is not.  `dm-integrity` is vulnerable to a partial rollback attack: it is possible to rollback parts of the disk without `dm-integrity` detecting it.  Therefore, `dm-integrity` is not (currently) sufficient for use with untrusted storage domains, which is a future goal of QubesOS.

**Comment by DemiMarie** [comment]
@tasket: what are your thoughts on using loop devices?  That’s my biggest worry regarding XFS+reflinks, which seems to otherwise be a very good choice for QubesOS.  Other approaches exist, of course; for instance, we could modify `blkback` to handle regular files as well as block devices.

**Comment by 0spinboson** [comment]
I really wish the FS's name wasn't a misogynistic slur. That aside, my only experience with it, under 4.0, had my Qubes installation become unbootable, and I found it very difficult to 'fix', relative to a system built on LVM. And that does strike as relevant to the question whether Qubes switches, which imo is only partly addressable via improving the documentation (since the other part is the software we have to use to restore).

**Comment by DemiMarie** [comment]
> FS's name wasn't a misogynistic slur

@0spinboson would you mind clarifying which filesystem you are referring to?

**Comment by tasket** [comment]
> Will it be possible to reserve space for use by discards? A user needs to be able to free up space even if they make a mistake and let the pool fill up.

Yes, its simple to allocate some space in a pool using a non-zero thin lv. Just reserve the lv name in the system, make it inactive, and check that it exists on startup.

Further, it would be easy to use existing space-monitoring components to also pause any VMs associated with a nearly-full pool and then show an alert dialog to the user.

> it is possible to rollback parts of the disk without `dm-integrity` detecting it.

I thought the journal mode would prevent that? I don't know it in detail, but something like a hash of the hashes of the last changed blocks, computed with the prior journal entry, would have to be in each journal entry.

> what are your thoughts on using loop devices? That’s my biggest worry regarding XFS+reflinks

I forgot they were a factor... its been so long since I've used Qubes in a file-backed mode. But this should be the same for Btrfs, I think.

FWIW, the XFS reflink suggestion was more speculative, along the lines of "What if we benchmark it for accessing disk images and its almost as fast as thin LVM?". The regular XFS vs Ext4 benchmarks I'm seeing suggest it might be possible. Its also not aligned with the Stratis concept, as that is closer to thin LVM with XFS just providing the top layer. (Obviously we can't use Stratis itself unless it supports a mode that accounts for the top layer being controlled by domUs.)

Also FWIW: XFS historically supported a 'subvolume' feature for accessing disk image files, instead of loopdevs. It requires certain IO sched conditions are met before it can be enabled.

**Comment by 0spinboson** [comment]
> > FS's name wasn't a misogynistic slur
> 
> @0spinboson would you mind clarifying which filesystem you are referring to?

'Butterface', was intentional, afaik.

**Comment by dmoerner** [comment]
Basic question: If I install R4.1 with BTRFS by selecting custom, and then using Anaconda to automatically create the Qubes partitions with BTRFS, is that sufficient for the default pool to use BTRFS-Reflink? Or do I have to do something extra for the "Reflink" part?

**Comment by rustybird** [comment]
> If I install R4.1 with BTRFS by selecting custom, and then using Anaconda to automatically create the Qubes partitions with BTRFS, is that sufficient for the default pool to use BTRFS-Reflink?

Yes

**Comment by noskb** [comment]
[ Change defaults: Btrfs should use 'dup' metadata on encrypted devices #319 ](https://github.com/kdave/btrfs-progs/issues/319)

**Comment by tasket** [comment]
I don't know if there's a separate issue for this, but possible Btrfs + fscrypt integration in Fedora seem relevant here:
> The system by default will be encrypted with an encryption key stored in the TPM and bound to the signatures used to sign the bootloader/kernel/initrd, 
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/LYUABL7F2DENO7YIYVD6TRYLWBMF2CFI/

**Comment by brendanhoar** [comment]
> I don't know if there's a separate issue for this, but possible Btrfs + fscrypt integration in Fedora seem relevant here:
> 
> The system by default will be encrypted with an encryption key stored in the TPM and bound to the signatures used to sign the bootloader/kernel/initrd

[If a separate ticket/forum discussion is opened for this, I will move this comment there.]

FWIW, I would want to avoid requiring TOM-stored data encryption keys by default, as it ties the user's data to system hardware that can fail.

The approach does makes some sense in an enterprise setting, primarily ensuring that data on storage devices separated from the machine with the TPM are provably unrecoverable during reuse/e-cycling. Business data is often backed up off enterprise endpoints by tools such as One Drive (e.g. often subsuming the Documents folder on windows in recent deployments), so they usually have the hardware failure risk mitigated via backups by default.

For non-enterprise users, esp. the user audience for Qubes, there should be flexibility in how keys are handled. Storage-subsystem-detached keys are useful for some, but the user must make the choices on privacy/security vs data-loss risk.

B

**Comment by DemiMarie** [comment]
Confirmed: [dm-thin is optimized for in-place overwrites of already-provisioned blocks, not breaking sharing or provisioning new blocks](https://lore.kernel.org/dm-devel/CAJ0trDZJW6uXkeCUCBRyJ_wg70wGmqSs+r+5HKACrKETN-Kdmg@mail.gmail.com/).

**Comment by tasket** [comment]
May also explain why the default metadata volume sizes seem insufficient.

Aside from performance, I'd have to say my experience with Btrfs has been more stable than with tLVM.  When something does go wrong with Btrfs, its easier to diagnose and recover.

There is also the problem of extra wear from write-amplification.  When I look at stats other people have posted for the nvme models I'm using, I'm seeing much higher rates of wear-out on my drives (that have had Qubes on tLVM).  Compared to people reporting a similar amount of lifetime read-access GB, my own drives are seeing > 3X the wear-indicator values.

**Comment by rustybird** [comment]
@kalkin There's a newer `file-reflink` storage driver that's used for the non-default Btrfs installation layout since R4.0.1.

**Comment by tlaurion** [comment]
@andrewdavidwong Release TBD means not planned for 4.2?  This ticket title should be updated for 4.3 and performances comparisons for default installs should be taken into consideration as well. It is to be noted that most OSes are moving away of TLVM for XFS/BRTFS.

Some history:
- Fedora 33 swtiched to BRTFS partitioning by default for workstations, having swap and rootfs in different LUKS containers and still default to BRTFS as Fedora 38.
- Ubuntu Lunar Lobster followed steps and decided to switch to BRTFS as well.
- OpenSuse followed step as well.
- Debian still relies on a thin LVM.

Also, ext4 as fixed inodes at time of formatting the partition, as opposed to XFS/BRTFS which are dynamic, which Qubes extends causing issues for some users.


I interested into knowing what Heads should support in the future for space constaints reasons and to prepare for changes. Also, Qubes is first class citizen, but not the only OS deployed. I was wondering what are the directions of QubesOS considering that BRTFS is default since Fedora33.

----

cross-posts linking to each other:
- https://forum.qubes-os.org/t/btrfs-and-qubes-os/6967/28?u=insurgo
- https://github.com/osresearch/heads/issues/1474

**Comment by DemiMarie** [comment]
@tlaurion BTRFS appears to have significant problems with I/O intensive workloads.  There are known problems that can result in unbounded latency spikes.

**Comment by tlaurion** [comment]
> @tlaurion BTRFS appears to have significant problems with I/O intensive workloads.  There are known problems that can result in unbounded latency spikes.

@DemiMarie those were documented? They should be referenced here. From my experience, the benefits are definitely overpowering TLVM

**Comment by andrewdavidwong** [comment]
> Release TBD means not planned for 4.2?

The "Release TBD" milestone no longer exists ([explanation](https://groups.google.com/g/qubes-users/c/gUZokSAAeM4)).

**Comment by DemiMarie** [comment]
> > @tlaurion BTRFS appears to have significant problems with I/O intensive workloads.  There are known problems that can result in unbounded latency spikes.
> 
> @DemiMarie those were documented? They should be referenced here. From my experience, the benefits are definitely overpowering TLVM. Care to share some I/O intensive workloads examples?

@marmarek did the benchmarks.  IIRC he found that BTRFS and XFS were not any faster than thin LVM in a workload (Qubes OS openQA tests, IIRC) that should have favored them.

**Comment by tlaurion** [comment]
> > > @tlaurion BTRFS appears to have significant problems with I/O intensive workloads.  There are known problems that can result in unbounded latency spikes.
> > 
> > @DemiMarie those were documented? They should be referenced here. From my experience, the benefits are definitely overpowering TLVM. Care to share some I/O intensive workloads examples?
> 
> @marmarek did the benchmarks.  IIRC he found that BTRFS and XFS were not any faster than thin LVM in a workload (Qubes OS openQA tests, IIRC) that should have favored them.

I'm surprised and looking forward to read the benchmarks @marmarek

**Comment by tlaurion** [comment]
@DemiMarie @marmarek 

It is also with noting that whatever benchmark that might have happened prior of kernel 6.1+ release might need to be redone, since whatever bottleneck on IO that affected the kernel prior has now vanished.


6.2 real life testing on database situation shows some interesting hypothesis/validation showing important boosts that needs to be read
- https://blog.pythian.com/btrfs-performance-compared-lvmext4-regards-database-workloads/

Also:
>According to Phoronix, Btrfs has enjoyed some performance improvements with Linux 6.x kernel after the major bug fixes. Some of the articles that report on these improvements are:

>- [Btrfs Brings Some Great Performance Improvements With Linux 6.1](^1^): This article claims that Btrfs has introduced async buffered writes and IO_uring support, which increased the direct IO limit for read and improved the throughput by 3x on a sample workload. It also mentions a significant FIEMAP speed improvement and a change to the algorithm for how extents are enumerated that leads to orders of magnitude speed boost.
>- [Btrfs Enjoys More Performance With Linux 6.3](^3^): This article states that Btrfs has added some performance optimizations and new features, such as big timestamps, FIEMAP speed improvement, and async buffered writes for compressed extents. It also reports some 3~10x speedups in some benchmarks.
>- [Btrfs In Linux 6.5 May Bring A Cumulative Performance Improvement For Metadata-Heavy Operations](^4^): This article says that Btrfs has brought a cumulative performance improvement for metadata-heavy operations, such as reading extent buffer in one-go, simplifying IO tracking and bio submission, and avoiding unnecessary reads in scrub code.

>These articles suggest that Btrfs has improved its performance by addressing some of the bugs and limitations that affected its previous versions. However, they also acknowledge that Btrfs still has some challenges and drawbacks, such as RAID56 issues, fragmentation, balance operations, and stability. Therefore, the performance of Btrfs may vary depending on the workload type, the disk layout, the compression algorithm, and the mount options. 😊

>Source: Conversation with Bing, 8/22/2023
>(1) Btrfs Brings Some Great Performance Improvements With Linux 6.1 - Phoronix. https://www.phoronix.com/news/Linux-6.1-Btrfs. 
>(2) Btrfs Enjoys More Performance With Linux 6.3 - Phoronix. https://www.phoronix.com/forums/forum/software/general-linux-open-source/1374178-btrfs-enjoys-more-performance-with-linux-6-3-including-some-3~10x-speedups.
>(3) Btrfs In Linux 6.5 May Bring A Cumulative Performance ... - Phoronix. https://www.phoronix.com/news/Btrfs-Linux-6.5.
>(4) Btrfs - Phoronix. https://www.phoronix.com/search/Btrfs.
>(5) undefined. https://btrfs.wiki.kernel.org/index.php/Main_Page.

**Comment by marmarek** [comment]
The previously mentioned tests (the lower time the better):

- [Test suite run on LVM](https://openqa.qubes-os.org/tests/73717): 1:01:00
- [Test suite run on btrfs](https://openqa.qubes-os.org/tests/73968): 1:44:00 and in one test VM failed to start within 90s timeout
- [Test suite run in XFS](https://openqa.qubes-os.org/tests/73996): 1:01:00

Both tests were on the same real laptop, and same software stack 
(besides the partitioning). Furthermore, in another run, fstrim / timed
out on btrfs after 2min (that was after installing updates in all
templates, so there probably was quite a bit to trim, but still, that's
a ~200GB SSD, so not that big and not that slow). Seeing this results,
I've rerun it several times, but got similar results.

Different test, much less heavy on I/O:

- [LVM](https://openqa.qubes-os.org/tests/73718): 0:34:49
- [btrfs](https://openqa.qubes-os.org/tests/73969): 0:33:27

So, at least not a huge regression, but not a significant improvement
either.

> It is also worth noting that whatever benchmark that might have happened prior of kernel 6.1+ release might need to be redone, since whatever bottleneck on IO that affected the kernel prior has now vanished and penalized brtfs vs lvm which improvements is stalling.

Those were running on kernel-latest at that time, so at least 6.3.x. But even then, if only the very latest kernel version would start to work fine, it's not enough for switching the default. The LTS version needs to work well for that.
In any case, it's way to late for this for R4.2. We may re-test, and re-consider for R4.3.

**Comment by tasket** [comment]
Thanks for the test results.  TLVM has outsized lag times with snapshot rotation for larger-sized volumes (>32GB data content), so that should be factored in as well.  This is probably why my subjective experience says that Btrfs feels faster than TLVM.

FWIW, `fstrim` should generally be avoided.  Its default discarding pattern is aggressive and not used by the filesystems themselves in discard mode for a reason, as it creates higher fragmentation and greater demand on metadata resources.  In my experience, fstrim use can also contribute to or trigger a TLVM [failure](https://github.com/QubesOS/qubes-issues/issues/3964).

I think stability is an overriding concern with choice of storage system, and I would put Btrfs head and shoulders above either XFS or TLVM in that regard.  My PC experience over the past decade has been mostly on Btrfs and (because of Qubes) TLVM, and the difference has become stark.

OTOH, I realize performance is a sensitive issue for Qubes, which has suffered a negative efficiency trend.  However, I'd also consider how that trend compares to the trend for PC systems in general when weighing performance vs. stability priorities.

**Comment by tasket** [comment]
BTW, even when you want to run `fstrim` manually for some maintenance objective and you reduce the granularity with `--minimum` (which does help), it is still hitting metadata resources with new deltas in a very short period of time, and the volume(s) might not be small.  Compare that to snapshot rotation, where deltas accumulate via routine user app use or system updates, and the sudden changes are mainly limited to removing deltas (freeing metadata space) when snapshots are removed.

All of the CoW/snapshot capable systems have issues with surges in metadata use.  NTFS is famous for gradually degrading in performance and becoming unbearably slow for minutes or even days.  Btrfs is a bit closer to NTFS in this, and I wonder if they have made a better trade-off for uptime vs performance.  ZFS is reported to degrade as the number of snapshots increase.  However, TLVM's issues haven't been tempered like these other systems; one gets strange inconsistencies that have to be diagnosed before applying a course of disjointed home remedies.

> What do you recommend instead? Should the filesystem be mounted with the discard option?

@DemiMarie Definitely use the mount option, which I think is standard practice now.  The mount option helps moderate fragmentation resulting from discards.  I think `fstrim` should remain something that is only invoked manually when an admin wants to address a special case (such as having forgotten to mount with _'discard'_).

Some recommendations:

The Btrfs _'nodesize'_ setting affects contention and thus latency for metadata-intensive operations, of which our fragmented VM image files require in abundance.  It is reasonable to assume the 16KB default is _not_ optimized for active disk image files.  Therefore, I suggest testing Qubes usage with _'nodesize'_ set to a smaller value like 8KB or 4KB.  This can be taken together with other optimizations:

* Btrfs _'no-holes'_ and _'skinny-metadata'_ options
* 4KB guest fs block size
* Use F2FS or XFS for guest private vols, both of which have considerably lower write amplification rates vs EXT4
* For guest root fs, compression should be considered

Some experimentation could be done beyond that, however.  To reduce the level of data fragmentation resulting from random writes (not just discards), ideally we would want to steak out a middle-ground between the 4KB minimum extent size that Btrfs normally uses and the 64KB (but usually >256KB) minimums that TLVM uses.  Also, active defragmentation and deduplication (the two often having opposite effects) should not be completely discounted; periodically running them both with extent-size thresholds that complement each other could improve overall efficiency and responsiveness.

**Comment by Rudd-O** [comment]
Would love to see a TLVM / btrfs / ZFS (ashift 12) performance comparison now that we have the three drivers in upstream.

**Comment by tlaurion** [comment]
> Would love to see a TLVM / btrfs / ZFS (ashift 12) performance comparison now that we have the three drivers in upstream.

Would love to see that!

| File system | Pros | Cons | References |
| ----------- | ---- | ---- | ---------- |
| TLVM        | - Offers more flexibility and scalability for creating and managing virtual disks for Qubes OS and the VMs. <br> - Uses thin provisioning, which means it only allocates the space that is actually used by the VMs, instead of reserving all the space in advance. | - Adds more complexity and overhead to the system, and may not work well with some file systems that do not support discard. <br> - Uses a lot of metadata to keep track of the changes in the data, especially when using features like snapshots, which can grow too large and slow down the system or cause errors. <br> - May be more prone to corruption if the system is hard reset. | [1], [2], [3], [4] |
| btrfs       | - Supports many advanced features that can improve the performance and reliability of Qubes OS and the VMs, such as compression, encryption, deduplication, checksums, snapshots, subvolumes, RAID, etc. <br> - Supports resizing and thin provisioning natively, which means it does not need LVM layer to work with Qubes OS. <br> - Has a good balance between large and small extent sizes, which can reduce fragmentation and metadata use. | - Has some issues with SELinux relabeling and root filesystem resizing, which may require some manual intervention or workaround to solve. <br> - Uses a lot of metadata and space to keep track of the changes in the data, which may require more frequent maintenance and cleanup to avoid performance degradation or errors. <br> - Is still relatively new and experimental compared to other file systems, which means it may have some unknown bugs or limitations that may cause problems. | [5], [6], [7], [8] |
| ZFS         | - Supports many advanced features that can improve the performance and reliability of Qubes OS and the VMs, such as compression, encryption, deduplication, checksums, snapshots, subvolumes, RAID, etc. <br> - Supports encryption natively, which means it does not need LUKS layer to work with Qubes OS. <br> - Has some unique features that can enhance the caching and logging capabilities of the system, such as ARC, SLOG, L2ARC, etc. <br> - Requires a custom kernel parameter (ashift) to match the disk block size, which can improve the performance and efficiency of ZFS by aligning the data with the physical structure of the disk. Ashift can also prevent wasting space on the disk by avoiding partial blocks. | - Uses a lot of memory and CPU resources to perform its operations, which may affect the performance and availability of other processes on the system. ZFS may also require more tuning and optimization to work well with Qubes OS and the VMs. <br> - Is still relatively new and experimental for Linux systems, which means it may have some compatibility and stability issues with Qubes OS and its features. ZFS may also have some bugs or limitations that may cause problems. | [9], [10], [11], [12] |


- [1]: TLVM: Thinly-provisioned Logical Volume Manager - https://www.kernel.org/doc/html/latest/admin-guide/device-mapper/thin-provisioning.html
- [2]: TLVM: A New Logical Volume Manager for Linux - https://www.usenix.org/system/files/conference/atc16/atc16-paper-lee.pdf
- [3]: TLVM: A New Logical Volume Manager for Linux - https://www.youtube.com/watch?v=0QZLw8nY4gk
- [4]: TLVM: How to recover from a hard reset? - https://github.com/QubesOS/qubes-issues/issues/2729
- [5]: btrfs: B-tree file system - https://btrfs.wiki.kernel.org/index.php/Main_Page
- [6]: btrfs: SELinux relabeling issue - https://github.com/QubesOS/qubes-issues/issues/1817
- [7]: btrfs: Performance tuning - https://btrfs.wiki.kernel.org/index.php/Performance_tuning
- [8]: btrfs: Root filesystem resizing issue - https://github.com/QubesOS/qubes-issues/issues/1120
- [9]: ZFS: Zettabyte File System - https://openzfs.github.io/openzfs-docs/
- [10]: ZFS: ashift parameter - https://openzfs.github.io/openzfs-docs/Basic%20Concepts/vdevs.html#ashift
- [11]: ZFS: Performance tuning - https://openzfs.github.io/openzfs-docs/Performance%20and%20Tuning/index.html
- [12]: ZFS: Compatibility and stability issues with Linux - https://github.com/openzfs/zfs/issues

**Comment by DemiMarie** [comment]
@tlaurion links 2 and 3 are 404, and links 4, 6, and 8 are completely irrelevant.  Also ZFS is not experimental; there are plenty of production workloads (like Let’s Encrypt’s main databases!) that use it.

**Comment by tlaurion** [comment]
> Also ZFS is not experimental; there are plenty of production workloads (like Let’s Encrypt’s main databases!) that use it.

@DemiMarie ashift and alignment to disk blocks is interesting property there, as well as dodging complexity of having LUKS enclosed BRTFS or current LUKS->TLVM->volumes complexity and failures. I'm not a FS expert, but I have suffered from TLVM qubes defaults in the past and really looking forward for a switch for whatever better. TLVM is known to fail the user, it's not a question of how, which is answered everywhere, but when. Your praises for BRTFS were proven by testing by the community, your call was heard by other projects, now the question is where we go to and what will be proposed to the user next. TLVM still in 4.3 or ZFS/BRTFS. 

Only testing under different use case scenarios will answer that question, where let's encrypt database scenario doesn't correspond to server Xen virtualization servers, which our use case more correspond to. **I wish we had some infra admins of the sort chiming in. That input would be significative **.

**Comment by andrewdavidwong** [comment]
@tlaurion: FWIW, I don't think it's very useful to post large quantities of ChatGPT output in qubes-issues, unless you're willing to say, at minimum, that you've personally vetted the output and agree with it.

**Comment by DemiMarie** [comment]
> @tlaurion: FWIW, I don't think it's very useful to post large quantities of ChatGPT output in qubes-issues, unless you're willing to say, at minimum, that you've personally vetted the output and agree with it.

Exactly.  Such vetting clearly was not done here, as shown by the broken and irrelevant links.

**Comment by Rudd-O** [comment]
Seconding Demi here.  ZFS is not experimental — the only reason it's not included in mainline is because it's not GPL, not because it's "waiting for the bugs to shake out".

**Comment by tasket** [comment]
I usually avoid ZFS for the same reasons Linus Torvalds does; I simply don't trust Oracle and large corporations are getting bolder with their open source IP rug-pulls (see Red Hat, not to mention Oracle's past attempt with Java APIs).

For people who do trust Oracle, I remain open-minded about the possibility of supporting OpenZFS in Wyng [backup](https://github.com/tasket/wyng-backup/issues/110).  But I also don't see a clear cut way to obtain image file metadata there; in that case Wyng only performs as well as a typical incremental backup.  Maybe someone could enlighten me, but AFAICT ZFS is not currently providing features that would enable efficient metadata-driven backups to _non-ZFS_ filesystems.

---

The idea Btrfs is "perpetually unfinished" doesn't ring true; the basic extent and file allocation format has been described as finished by the Btrfs devs, while they also state upfront that changes are still being made.  None of the popular OS-running filesystems have been finalized – see XFS as a particularly old example where significant features are still being added.

The Btrfs format has some amazing qualities, for example the ability to describe a volume (or sub-volume) in a way that can replicate deduplication/reflinks of file data on a different Btrfs filesystem (e.g. `btrfs-send` knows what data is shared by reflinked files).  The reason why is that everything a file/inode directly references in Btrfs is considered a _logical_ entity, so its easy to transfer while retaining its original identity (and when files reference them more than once, it follows that is easy to replicate as well).  Very promising stuff, IMO.

Finally, versus thin-lvm, Btrfs has been a quantum leap in reliability for me; I've used and provided support for both for as long as each has had a Qubes driver (longer than that for Btrfs) and there is no question in my mind about the difference.

**Comment by rustybird** [comment]
One exciting feature coming up in Btrfs is fscrypt support ("file-based encryption"). The reason I find it exciting in a Qubes OS context is because they're making it extent-based, i.e. not just individual files but individual fragments of files can have their own encryption keys. Reading the tea leaves on their mailing list, I get the impression that this design will - eventually, but probably not from day one - allow reflinking an unencrypted (or differently encrypted) source file into a destination file where subsequent writes will then be encrypted with a new key. Which would be exactly what is needed to support ephemeral `snap_on_start` volumes, solving https://github.com/QubesOS/qubes-issues/issues/904#issuecomment-1929766110.

**Comment by tasket** [comment]
I think the facts show that the fate of OpenJDK and derivatives (what most mobile devices rely on) was left to the US courts – which are now very much in a mood to issue new rulings on supposedly settled case law.

And CentOS was effectively disbanded. The replacements don't have access to RH "proprietary" fixes and modifications that they distribute to RHEL customers (a "new fact" about software that the legal world seems comfortable with).  If corporations are ready to attempt de-facto nullification of the GPL then licensing is a top concern.

**Comment by tlaurion** [comment]
> The previously mentioned tests (the lower time the better):
> 
> - [Test suite run on LVM](https://openqa.qubes-os.org/tests/73717): 1:01:00
> - [Test suite run on btrfs](https://openqa.qubes-os.org/tests/73968): 1:44:00 and in one test VM failed to start within 90s timeout
> - [Test suite run in XFS](https://openqa.qubes-os.org/tests/73996): 1:01:00
> 
> Both tests were on the same real laptop, and same software stack 
> (besides the partitioning). Furthermore, in another run, fstrim / timed
> out on btrfs after 2min (that was after installing updates in all
> templates, so there probably was quite a bit to trim, but still, that's
> a ~200GB SSD, so not that big and not that slow). Seeing this results,
> I've rerun it several times, but got similar results.
> 
> Different test, much less heavy on I/O:
> 
> - [LVM](https://openqa.qubes-os.org/tests/73718): 0:34:49
> - [btrfs](https://openqa.qubes-os.org/tests/73969): 0:33:27
> 
> So, at least not a huge regression, but not a significant improvement
> either.
> 
> > It is also worth noting that whatever benchmark that might have happened prior of kernel 6.1+ release might need to be redone, since whatever bottleneck on IO that affected the kernel prior has now vanished and penalized brtfs vs lvm which improvements is stalling.
> 
> Those were running on kernel-latest at that time, so at least 6.3.x. But even then, if only the very latest kernel version would start to work fine, it's not enough for switching the default. The LTS version needs to work well for that.
> In any case, it's way to late for this for R4.2. We may re-test, and re-consider for R4.3.

I would love to fucus on currently referred testing results as being the base for not going forward with BRTFS.

Maybe OP should refer to those test results, kernel versions to state clearly what could be the causes of discrepancies between qubesos forum post showing better performance gains simply by switching to brtfs from thin lvm?

@tasket suggested some optimizations on filesystems options to boost brtfs performance for better performance under qubesos.

@Rudd-O made really clear of performance and stability gains of openzfs vs TLVM and brtfs, where licensing still is an issue, even though Ubuntu dodged the problem altogether a while ago and never got sued.


My past attempts here were to recap on the current states of improvements of brtfs. This thread stated possible causes of performance degradation of TLVM over brtfs for large volumes snapshot rotation, trimming etc.

It's currently hard to grasp what is the current state of things and how to get things forward.

The reason I insist on this is because downstream projects insist on cloning templates and diverging through salt recipes, which TLVM cannot support since no dedup is possible. Brtfs supports offline dedup, which bees permits to accomplish if deployed so that backups can be restored without filling disks. Openzfs would fit the need even better by compressing and offering online deduplication, resolving space consomption extend life of SSD drives, which otherwise as been claimed again and again that qubesos over other OSes is a drive life shortener.

Users wants faster vm boot/shutdown time, proper backup/restore mechanisms, hardware life and battery efficiency.

Openzfs is known to be a ram hugger for dedup. TLVM overhead is known, large disks qubes show that TLVM loses over brtfs, openzfs shines in all aspect but licence. Where to go next and how?

What is the current state of the art with current versions and where/how to get this forward?
From what I get, brtfs proper filesystem creation and runtime options need to be defined properly to be compared correctly against current TLVM ones. Openzfs, if I understand well the situation, won't be part of the installer until Fedora does a move like Ubuntu did, otherwise QubesOS won't take the decision to do what Ubuntu did by themselves.

So it leaves us to brtfs. So how to show without any doubt that brtfs can be a better candidate then TLVM (which most other OSes dodged being default deployments)?

If QubesOS stays with Fedora as dom0, which is not panned to change anytime soon unless I missed something, then how to recuse SSD wear, improve large vm shutdown time and embrace templates space deduplication from cloning/diverging without switching to brtfs is openzfs is out of scope?

**Comment by DemiMarie** [comment]
> @DemiMarie :
> Maybe OP should refer to those test results, kernel versions to state clearly what could be the causes of discrepancies between qubesos forum post showing better performance gains simply by switching to brtfs from thin lvm?

Would it be possible to re-run benchmarks now?  I’d rather refer to up-to-date benchmarks than ones known to be stale.

**Comment by DemiMarie** [comment]
> Canonical has shown that this is a solved problem. We should do what they did and get on with other pressing issues such as Wayland, seL4, etc.

Wayland definitely needs to be implemented, and I’m going to be talking about GPU acceleration (which it will enable) at Xen Project Summit 2024.  Right now, seL4 [doesn’t provide sufficient protection against CPU vulnerabilities](https://github.com/seL4/seL4/issues/1108), so it isn’t an option.

**Comment by tlaurion** [comment]
> > @DemiMarie :
> > Maybe OP should refer to those test results, kernel versions to state clearly what could be the causes of discrepancies between qubesos forum post showing better performance gains simply by switching to brtfs from thin lvm?
> 
> Would it be possible to re-run benchmarks now?  I’d rather refer to up-to-date benchmarks than ones known to be stale.

@marmarek i guess this question from @DemiMarie was addressed to you for open-qa tests

@DemiMarie meanwhile, there is no cost at editing OP with past results so all people getting here are at least clear on the why's and versions related to brtfs having not been considered in the past because bad press, no?

**Comment by tlaurion** [comment]
On my side, I postponed building https://github.com/tlaurion/qubes-bees directly from qubes-builder v2 (which didn;t work last time I attempted that goal) to build the rpm directly under fedora-37 to produce an installable bees rpm. Will update https://forum.qubes-os.org/t/bees-and-brtfs-deduplication/20526/6 later on with some meat to chew on when I have it.

Testing to see and compare gains of deduped space with/without bees, as a start.

**For those interested to test dedup gains on test machines only, here is the produced RPM (gzipped because github constraints).**: [bees-0.10-1.fc37.x86_64.rpm.gz](https://github.com/QubesOS/qubes-issues/files/14972516/bees-0.10-1.fc37.x86_64.rpm.gz)

**Comment by tlaurion** [comment]
Some preliminary not so convincing bees results on x230 https://forum.qubes-os.org/t/bees-and-brtfs-deduplication/20526/10

**Comment by kocmo** [comment]
> Additionally, LVM thin pools do not support checksums. This can be achieved via dm-integrity, but that does not support TRIM.

Ext4 has metadata checksums enabled since e2fsprogs 1.43, so at least some filesystem integrity checking is happening inside VMs:

```
root@sys-firewall /h/user# dumpe2fs /dev/mapper/dmroot | grep metadata_csum
dumpe2fs 1.47.0 (5-Feb-2023)
Filesystem features:      has_journal ext_attr resize_inode dir_index orphan_file filetype needs_recovery extent 64bit flex_bg metadata_csum_seed sparse_super large_file huge_file dir_nlink extra_isize **metadata_csum** orphan_present
```

Does Qubes have mechanisms to report kernel errors from VMs and dom0 to the user, via toast notifications or so?

In Qubes 4.2.1, 4.2.2 dom0 systemd journal continuously gets repeated PAM error messages :-/
```
dom0 pkexec[141170]: PAM unable to dlopen(/usr/lib64/security/pam_sss.so): /usr/lib64/security/pam_sss.so: cannot open shared object file: No such file or directory
dom0 pkexec[141170]: PAM adding faulty module: /usr/lib64/security/pam_sss.so
```

**Comment by tlaurion** [comment]
@DemiMarie https://github.com/tasket/wyng-backup/issues/211

**Comment by DemiMarie** [comment]
@tlaurion Can you provide proper benchmarks?  Last benchmarks by @marmarek found that BTRFS was not faster than LVM2, which is why LVM2 is still the default.

**Comment by tlaurion** [comment]
> @tlaurion Can you provide proper benchmarks?  Last benchmarks by @marmarek found that BTRFS was not faster than LVM2, which is why LVM2 is still the default.

@DemiMarie unfortunately I don't. Notes are scattered under bees and wyng-backup issues for the moment which is how I optimized my btrfs setup and would never look back to thinlvm ever again, until zfs is figured out to be added under installer. 

But prior of being to do a proper perf comparison, defaults of btrfs filesystem creation and fstab mount options need revisiting, including détection of Block device type being HDD/(ssd/nvme).

Incomplete notes out of my head:
- trimming needs to be async otherwise qubes snapshot rotation kills IO waiting on btrfs-transaction putting system into apparent frozen state. 
- btrfs : only system needs to be safeguarded with DUP, otherwise incredible overhead. Metadata and data should be single instead of DUP (block device type dependent) 
- mode optimization should be specified under fstab (block type dépendant)

Will try to revisit scattered issues and post them here with further edit when I have a bit more time to invest in this issue. Or at least point this comments to those comments.

**Comment by marmarek** [comment]
> until zfs is figured out to be added under installer.

Very unlikely (unless available in upstream kernel, which is also very unlikely). We have CI job that checks if ZFS pool works, and quite often we find that it doesn't work with the latest kernel yet. So official ZFS support would hold back kernel updates, which I heard from @DemiMarie is completely unacceptable if doing any sort of GPU acceleration work (which we do want at some point). Currently the said CI job doesn't work, because [there is no dkms package for Fedora 41](https://github.com/openzfs/zfs/issues/16590) (which R4.3 is based on) yet.

Back to topic:
> trimming needs to be async otherwise qubes snapshot rotation kills IO waiting on btrfs-transaction putting system into apparent frozen state.

Is there any impact on resilience for power failure cases?

> mode optimizations should be specified under fstab (block type dépendant)

What do you mean?

Can you collect specific options that need to be set (fstab and elsewhere)?

**Comment by DemiMarie** [comment]
> > until zfs is figured out to be added under installer.
> 
> Very unlikely (unless available in upstream kernel, which is also very unlikely). We have CI job that checks if ZFS pool works, and quite often we find that it doesn't work with the latest kernel yet. So official ZFS support would hold back kernel updates, which I heard from @DemiMarie is completely unacceptable if doing any sort of GPU acceleration work (which we do want at some point). Currently the said CI job doesn't work, because [there is no dkms package for Fedora 41](https://github.com/openzfs/zfs/issues/16590) (which R4.3 is based on) yet.

For clarification: I am not sure if Qubes OS will use the latest stable kernel or the latest LTS kernel, and if Qubes OS will skip the first few releases in a stable branch.  If Qubes OS _will_ skip the first few releases (as seems likely, since these releases often have easily-found bugs), this might give ZFS sufficient time to catch up.

What _must_ be taken regularly for GPU acceleration are weekly updates in the upstream branch that Qubes OS has chosen to follow.  I would be highly surprised if those break OpenZFS, though there are obviously no guarantees.

**Comment by tlaurion** [comment]
> > trimming needs to be async otherwise qubes snapshot rotation kills IO waiting on btrfs-transaction putting system into apparent frozen state.
> 
> Is there any impact on resilience for power failure cases?

Not that i'm aware of; DUP is for HDD where ssd/nvme does its own thing.
Also note that for wyng-backups (my goal), I disabled volumes to keep globally as well, so impacts IO performance (I rely on single snapshots from wyng last backup)

> > mode optimizations should be specified under fstab (block type dépendant)
> 
> What do you mean?
> 
> Can you collect specific options that need to be set (fstab and elsewhere)?

@marmarek @tasket  current fstab:
`UUID=a8b2c51e-f325-4647-8fc2-bc7e93f49645 		/                       btrfs   subvol=root,x-systemd.device-timeout=0,noatime,compress=zstd,ssd,space_cache=v2,discard=async,noautodefrag 0 0
`
mainly for ssd:
- ssd, space_cache=v2 (default now), discard=async
and for reflink perf improvements:
- noautodefrag

One would avocate that system should be in DUP for resilience, but I have observed no impact:
```
[user@dom0 ~]
$ sudo btrfs fi usage /
Overall:
    Device size:		   1.79TiB
    Device allocated:		   1.59TiB
    Device unallocated:		 207.99GiB
    Device missing:		     0.00B
    Device slack:		     0.00B
    Used:			 553.73GiB
    Free (estimated):		   1.24TiB	(min: 1.24TiB)
    Free (statfs, df):		   1.24TiB
    Data ratio:			      1.00
    Metadata ratio:		      1.00
    Global reserve:		 512.00MiB	(used: 0.00B)
    Multiple profiles:		        no

Data,single: Size:1.57TiB, Used:549.96GiB (34.14%)
   /dev/mapper/luks-09f96e74-eb7e-4058-b3f4-e4f0ca8492fd	   1.57TiB

Metadata,single: Size:19.00GiB, Used:3.76GiB (19.81%)
   /dev/mapper/luks-09f96e74-eb7e-4058-b3f4-e4f0ca8492fd	  19.00GiB

System,single: Size:32.00MiB, Used:224.00KiB (0.68%)
   /dev/mapper/luks-09f96e74-eb7e-4058-b3f4-e4f0ca8492fd	  32.00MiB

Unallocated:
   /dev/mapper/luks-09f96e74-eb7e-4058-b3f4-e4f0ca8492fd	 207.99GiB
```

**Comment by rustybird** [comment]
@tlaurion:

I'm intrigued but skeptical that `mkfs.btrfs --metadata=dup` (the default) would have a notable impact on reflinking. Is there a benchmark showing this?

`--data=single` is already the default.

For mount options:

> ssd

Automatically applied if the drive's `/sys/block/foo/queue/rotational` attribute (which is passed through by dm-crypt) is `0`

> discard=async

Default on modern kernels, but unfortunately it's overridden to `default[=sync]` in fstab by the Anaconda installer since R4.2

>  noautodefrag

Default

**Comment by tlaurion** [comment]
> @tlaurion:
> 
> I'm intrigued but skeptical that `mkfs.btrfs --metadata=dup` (the default) would have a notable negative impact on reflinking. Is there a benchmark showing this?
> 
> `--data=single` is already the default.
> 
> For mount options:
> 
> > ssd
> 
> Automatically applied if the drive's `/sys/block/foo/queue/rotational` attribute (which is passed through by dm-crypt) is `0`
> 
> > discard=async
> 
> Default on modern kernels, but unfortunately it's overridden to `discard[=sync]` in fstab by the Anaconda installer since R4.2
> 
> >  noautodefrag
> 
> Default

By lack of proper benchmarking, I guess the culprit to fix here is then noautodefrag and redo testing

**Comment by rustybird** [comment]
@tlaurion:

> > `--data=single` is already the default.
> 
> Mine was dup on default install.

Odd. Does your filesystem span multiple block devices? Even then I don't see why that would result in `--data=dup`, unless Anaconda is doing something weird.

> I guess the culprit to fix here is then noautodefrag and redo testing

Just don't explicitly pass `autodefrag`. That's definitely unsuitable for file-reflink pools, because defragmentation duplicates shared data (so it's something that has to be done manually and selectively).

**Comment by tlaurion** [comment]
> It's unfortunate to watch so many cycles being spent to deal with issues that wouldn't even merit mention with ZFS.
> 
> Inability to use bleeding-edge kernel releases seems a bit disingenous as a reason to disqualify ZFS given how far back from the edge Qubes (rightfully!) stays.
> 
> Regarding distributing binaries (in the installer or otherwise), DKMS seems like it would be one good solution.

I tend to agree with that statement. Mixing dgpu pass-through with pool efficient management (online dedup being a big win for my use case, vs not caring at all with all graphic acceleration) woukd resolve most of my issues and time spent with bees which will keep doing of file dedup and consume CPU cycles I would prefer not spending.

Zfs > btrfs on all levels, once again.

**Comment by DemiMarie** [comment]
> > Inability to use bleeding-edge kernel releases seems a bit disingenous as a reason to disqualify ZFS given how far back from the edge Qubes (rightfully!) stays.
> 
> I don't want Qubes users to be in this situation: [openzfs/zfs#16590 (comment)](https://github.com/openzfs/zfs/issues/16590#issuecomment-2444970192) (6.11 kernel is in Qubes stable repo for 2+ weeks already).

To elaborate: Users who turn on GPU acceleration will need to update their kernel weekly to the latest release on their branch of choice.  For those who are using `kernel-latest`, this means that they will not be able to receive security patches for GPU drives.

**Comment by DemiMarie** [comment]
> > > Inability to use bleeding-edge kernel releases seems a bit disingenous as a reason to disqualify ZFS given how far back from the edge Qubes (rightfully!) stays.
> > 
> > 
> > I don't want Qubes users to be in this situation: [openzfs/zfs#16590 (comment)](https://github.com/openzfs/zfs/issues/16590#issuecomment-2444970192) (6.11 kernel is in Qubes stable repo for 2+ weeks already).
> 
> I completely agree, and I don't want them to either. That said, this was caused by a user adding a dependency outside of the project's control upon which the base system relied. If Qubes included it instead, presumably you'd hold back the kernel until there was a compatible ZFS version, just like you do for other Qubes-specific dependencies.

Holding back kernel updates is 100% incompatible with GPU acceleration, because Linux does not reliably issue security advisories for GPU driver vulnerabilities and so Qubes OS’s security team does not know which patches need to be backported.  Qubes OS will be offering GPU acceleration in the future because there are many users who simply cannot use Qubes OS without it.  Therefore, holding back kernel updates is not a sustainable solution.

It _might_ be possible to provide a ZFS DKMS package that _only_ supported LTS kernels, which is what Qubes OS ships by default.  However, there are users who must use `kernel-latest` to get support for their hardware and such users would not be able to use ZFS.

> The fact that users are going to such lengths to use ZFS shows there is demand for it (and @Rudd-O has already gone to great effort to lay some of the groundwork).

There is absolutely demand for ZFS, and for very good reason: ZFS is the most reliable filesystem available today, on any operating system.  To be clear, if I had a production server using an LTS kernel I would most likely choose ZFS for data volumes (though probably not the root volume unless I was on Ubuntu).  Qubes OS, however, is not a server operating system, and cannot rely on LTS kernels because it needs

> What ZFS brings to the table meshes very well with how Qubes aims to achieve its goals. I still cannot understand why we are dismissing it out of hand. (Its pedigree is also far superior to that of Btrfs.)

ZFS is _not_ being dismissed out of hand.  There _are_, however, multiple severe problems with it:

1. It is unclear whether ZFS binary kernel modules can legally be redistributed.  Redistribution in source is possible, but without reproducible DKMS builds it is incompatible with module signing for secure boot [^1] and causes other practical problems.
2. ZFS is an out of tree kernel module and therefore gets periodically broken by kernel updates.  For the reasons mentioned above holding back kernel updates will be a security no-no in the future.  LTS updates are much less likely to break ZFS, but not everyone uses LTS kernels.

[^1]: Reproducible builds solve this problem because the user can produce a module that is identical to the module that Qubes OS built.  This means that the signature can be extracted from the signed module, shipped alongside the source, and added back to the newly built kernel module to produce one that is validly signed.

**Comment by marmarek** [comment]
People can already use ZFS on Qubes OS, it isn't even that complicated to enable. But due to the above issues, it won't be part of the default installation.

**Comment by tlaurion** [comment]
Is there still desire to make btrfs first candidate of qubesos and fixing default fs/fstab options?

**Comment by tlaurion** [comment]
> @tlaurion:
> 
> > Is there still desire to make btrfs first candidate of qubesos and fixing default fs/fstab options?
> 
> The latter part can be done outside of this ticket, because it affects Btrfs users even while Btrfs isn't the default installation layout.
> 
> (If you find a way to reproduce the installer bizarrely setting up dup data on your single disk setup, please open an issue! 

I confirm that on fresh installation:
- Data is single
- Metadata is DUP (this is the problematic one if clones extensively used, should be single)
- System is DUP

> Otherwise, the only thing I'm aware of that needs some tweaking in the installer and in qubes-dist-upgrade is not to hardcode `discard[=sync]` in fstab.)

Agreed. Will apply this on test laptop (w530, quad core no HT, SSD Samsung 860 1TB) and see if adding `discard=async` alone under fstab (without touching metadata DUP or anything else) is enough to not explode pool with qusal + my usage of qubes+templates extensive clones will could issues and report back later on.

**Comment by tlaurion** [comment]
> @tlaurion:
> 
> > Is there still desire to make btrfs first candidate of qubesos and fixing default fs/fstab options?
> 
> The latter part can be done outside of this ticket, because it affects Btrfs users even while Btrfs isn't the default installation layout.
> 
> (If you find a way to reproduce the installer bizarrely setting up dup data on your single disk setup, please open an issue! Otherwise, the only thing I'm aware of that needs some tweaking in the installer and in qubes-dist-upgrade is not to hardcode `discard[=sync]` in fstab.)

Someone can launch perf comparison with that fstab setting changed vs default?

**Comment by DemiMarie** [comment]
> Someone can launch perf comparison with that fstab setting changed vs default on large disks and with multiple clones + snapshots of vm startup time? Total boot time to ready to work state?

Can you do that?

**Comment by tlaurion** [comment]
> > Someone can launch perf comparison with that fstab setting changed vs default on large disks and with multiple clones + snapshots of vm startup time? Total boot time to ready to work state?
> 
> Can you do that?

I thought this is what we waited to have to redo test bench, with newer kernel version etc. No, I do not have test bench to produce tests results that would have the desired impact. 

@marmarek I suggest redoing tests as of last time for comparison, whatever they were, without synced discards in fstab.

**Comment by tlaurion** [comment]
At https://github.com/QubesOS/qubes-issues/issues/6476#issuecomment-1689640103

@marmarek said
> We may re-test, and re-consider for R4.3.

**Comment by tlaurion** [comment]
I think discaed=sync is artifact of lvm being generalized to btrfs.

I tried both async and no discard at all, seeing no important loss on nvme but haven't tested on older hardware with ssd, but to be noted that  on newer hardware with so many cores without hyperthreading activated, host has a lot of CPU cycles to waste and async just pushes changes  down queues that are not even hitting iowait on newer hardware.


My recommendation here would be to either
- revert discard on 4.3 installer (both lvm/btrfs, rely on cron to do it when user leave machines up long enough)
- implement discard=async upon detection of nvme which implies newer hardware, if that's a good assumption
- at least change 4.3 iso installer default from synced discard to async and postpone the rest to later, since all observations point that synced discard is shooting ourselves in the foot.

Once one of those is chosen, relaunch tests of lvm vs btrfs. Post results here.


Bees has launched a call for testing where they finally changed their default crawling mode for vm use case with multiple snapshots. Cannot guarantee I will put community time on this anytime soon, but since it would be scratching my own itches, and Heads space is limited, and that Heads cannot have tools to support all filesystem that exist, I would need to prioritize tghis accordingly. I would not look back to lvm personally, but repeat that btrfs default installer choices need to be revisited, but first step is to prove one is better then the other and move this forward. I guess we are close of 4.3 feature freeze here, so its now or 4.3.1? We also know, from history, that iso installer defaults can be revisited but it's complicated when it comes to pool settings.

Some more input :
- btrfs varlibqubes would benefit being a subvolume of rootfs (dom0) for everything wyng /bees if there is still interest into that. Reasoning is simple: wyng needs to snapshot the pool/subvolume. 
  - Would be better if varlibqubes was independent, at least as a subvolume level. @tasket advocates into putting dom0 as a pool disk image, or seperate it so that dom0 can be backup easily, which is not the case right now.     - Note that /boot backup is resolved on wyng side. Lvm has advantage there, since dom0 can be snapshot and backup where btrfs can't as of now. 


@marmarek :
- Can discard=async be part of a iso build, and tested to compare performance vs lvm and we go from there?

**Comment by tasket** [comment]
I think either way, even if Btrfs doesn't become the default, there should be further accommodation of Btrfs (perhaps in the driver) in the form of:
* Creates/manages Qubes pools as their own subvolumes: This matters for Wyng, because transaction activity on the subvolume can remain high if that subvol also includes things like dom0 root. Metadata scans must be aborted/re-tried if fs transaction ids change during the scan.  Currently Btrfs users must manually run a script to convert the pool's directory into a subvolume.
* Periodic defragmentation of volume img files:  Some type of defragmentation is necessary on Btrfs when very large files receive many random updates, basically what happens to VM disk images.  I think the ideal place to initiate this is from the Qubes storage admin code, which could run `btrfs filesystem defrag` in the background on VM shutdown but only about once per week; running it online this way is effective and very _unobtrusive_.  This periodic approach avoids some of the write amplification and performance degradation that would occur from constant defrag via the Btrfs 'autodefrag' mount option.

I don't know if these need their own issues.

**Comment by DemiMarie** [comment]
Bad news: FIEMAP is only for debugging.  See <https://lore.kernel.org/all/aNGH9s0xoIg9Isk5@infradead.org/>.
